Dockerfile: use TARGETPLATFORM to build Docker

crazy-max opened the pull request:
follow-up #43529
~needs https://github.com/moby/moby/pull/44513~

fixes https://github.com/moby/moby/pull/41590
fixes https://github.com/moby/moby/issues/41862
fixes https://github.com/moby/moby/issues/41519
fixes https://github.com/moby/moby/issues/39380
closes https://github.com/moby/moby/pull/43492
closes https://github.com/moby/moby/pull/43613

Better support for cross compilation so we can fully rely on `--platform` flag of buildx for a seamless integration.

This removes unnecessary extra cross logic in the Dockerfile, removes `DOCKER_CROSSPLATFORMS` and `CROSS` vars and some hack scripts as well.

Non-sandboxed build invocation is still supported and dev stages in the Dockerfile have been updated accordingly.

Bake definition and GitHub Actions workflows have been updated accordingly as well.

cc @thaJeztah @rumpl @vvoland

crazy-max committed:
docs: contributing docs update for cross comp

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah reviewed (commented):
was still reviewing, but had one comment pending already; let me submit that one

crazy-max commented on Dockerfile:
```diff
@@ -540,115 +523,85 @@ RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
 ARG PLATFORM
-ENV PLATFORM=${PLATFORM}
 ARG PRODUCT
-ENV PRODUCT=${PRODUCT}
```
Doesn't look like it as dev image does not use the `build` stage anyway. `make shell` will run the dev image and pass env vars through https://github.com/moby/moby/blob/3e39ec60dab859f70577ea8561fe5bda5236749e/Makefile#L33-L79

thaJeztah replied:
Hm.. I wonder if we changed things on that at some point; you're correct; doing `make binary shell` can produce a different `PRODUCT` than using `./hack/make.sh binary` inside the dev container.

I guess that's ok (don't _have_ to bake the info into the image)

corhere commented on hack/make/.binary:
```diff
@@ -67,30 +74,66 @@ source "${MAKEDIR}/.go-autogen"
+fi
 
-echo "Created binary: $DEST/$BINARY_FULLNAME"
+if [ -n "$DOCKER_STRIP" ]; then
+	DOCKER_LDFLAGS+=" -s"
+fi
```
This looks new. What does it have to do with cross-compilation?

corhere commented on hack/make/.mkwinres:
```diff
@@ -9,14 +9,19 @@ quadVersionNum() {
 	fi
 }
 
+if [ ! -x "$(command -v go-winres)" ]; then
+	echo >&2 "go-winres not found, skipping manifesting binary"
+	exit 0
+fi
```
dockerd.exe _has to_ be manifested or it will be subtly broken. Windows APIs will lie to dockerd unless it has a manifest declaring it supports Windows 10. Until we commit pregenerated .syso files containing stub manifests, a missing `go-winres` needs to be a fatal build error.

corhere commented on hack/make/dynbinary-daemon:
```diff
@@ -4,13 +4,10 @@ set -e
+	fi
+	export DOCKER_STATIC=0
+	export PACKAGE=./cmd/dockerd
+	. hack/make/.binary
```
```suggestion
	source "${MAKEDIR}/.binary"
```
Being able to run independently of $PWD is a nice feature to retain.

corhere commented on hack/make/binary-proxy:
```diff
@@ -1,12 +1,13 @@
-
 set -e
 
+[ -z "$KEEPDEST" ] && rm -rf "$DEST"
```
What does this new footgun have to do with using TARGETPLATFORM?

corhere commented on hack/make/dynbinary:
```diff
@@ -1,11 +1,13 @@
 (
-	DEST="${DEST}-daemon"
-	ABS_DEST="${ABS_DEST}-daemon"
+	export KEEPDEST=1
```
What's the point of `KEEPDEST` if you're setting it to 1 everywhere?

corhere reviewed (dismissed):
I didn't get past the first commit. There are so many behaviour changes to the `hack/make` scripts which have nothing to do with the PR description _or commit message_.

thaJeztah commented on hack/make.sh:
```diff
@@ -83,53 +82,58 @@ add_buildtag() {
+	DOCKER_BUILDTAGS+=" netgo osusergo static_build"
 fi
 
+if [ "$CGO_ENABLED" = "1" ] && [ "$(go env GOOS)" != "windows" ] && [ "$DOCKER_STATIC" != "1" ]; then
+	# pkcs11 cannot be compiled statically if CGO is enabled (and glibc is used)
+	DOCKER_BUILDTAGS+=" pkcs11"
+fi
```
This may have been from when the CLI was still in this repo.

But looks like still needed for containerd; see https://github.com/moby/moby/pull/44513#discussion_r1034626488

```
#22 [containerd-build 3/3] RUN --mount=from=containerd-src,src=/usr/src/containerd,rw     --mount=type=cache,target=/root/.cache <<EOT (set -e...)
#22 0.862 + bin/ctr
#22 1.921 + bin/containerd
#22 3.631 # github.com/containerd/containerd/cmd/containerd
#22 3.631 /usr/bin/ld: /tmp/go-link-508382094/000019.o: in function `New':
#22 3.631 /go/src/github.com/containerd/containerd/vendor/github.com/miekg/pkcs11/pkcs11.go:77: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

corhere replied:
Huh??

> This may have been from when the CLI was still in this repo.

The string `pkcs11` does not appear _anywhere_ in the removed lines on this PR diff. _This is brand new code._ There is no `//go:build` or `//+build` constraint which references the `pkcs11` tag anywhere in the repo, vendor tree included.

> But looks like still needed for containerd

_make.sh isn't in the loop for building containerd_, nor is `$DOCKER_BUILDTAGS` respected by the containerd build in `Dockerfile`. And I couldn't find a build constraint which references the `pkcs11` tag anywhere in github.com/containerd/containerd, either.

thaJeztah replied:
> The string pkcs11 does not appear anywhere in the removed lines on this PR diff. This is brand new code.

Ah! hm, yes, in that case "I don't know" (I saw your comment, and expect it was code copied/moved from one of the `hack/` scripts).


> make.sh isn't in the loop for building containerd, nor is $DOCKER_BUILDTAGS respected by the containerd build in Dockerfile. And I couldn't find a build constraint which references the pkcs11 tag anywhere in github.com/containerd/containerd, either.

Yeah, we won't need it in `hack/make.sh` for that, but we should check the containerd builds; the fact it isn't in their code repository could be an issue (their static build targets were broken various times, and still may be, e.g. https://github.com/containerd/containerd/issues/5824)

corhere replied:
No files in either github.com/moby/moby or github.com/containerd/containerd contain lines matching `^//go:build .*pkcs11` or `^//+build .*pkcs11`. What else is there to check?

Furthermore, the premise that the `hack/make.sh` code adding the `pkcs11` build tag is a well-intentioned attempt to fix static containerd builds makes no sense: the `pkcs11` tag is _added_ on _dynamically-linked_ builds.

corhere replied:
Mystery solved: [the build tag is referenced in the github.com/docker/cli build scripts.](https://github.com/docker/cli/blob/850fb6921cf7a208fd47d83074f27d0eb7ae268e/scripts/build/binary#L10-L12)

crazy-max committed:
bake: define DESTDIR

Keep the same output dir format in the bake definition
as the one used in make scripts.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: rename dev stages

Also remove DEV_IMAGE bake var that is not used

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
ci: add extra steps to check artifacts

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
ci: use GITHUB_REF and GITHUB_SHA to set version and commit

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
> I didn't get past the first commit. There are so many behaviour changes to the `hack/make` scripts which have nothing to do with the PR description _or commit message_.

@corhere Yes this was indeed ahead of the expectations, sorry about that. I will open a follow-up to fix issues with incorrect linking for docker-proxy and dockerd we currently encounter. It should be easier to review now.

crazy-max committed:
IAmStatic not used anymore

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
hack: use PKG_CONFIG var when checking libdevmapper

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
hack: do not set buildmode pie for statically linked binary

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
hack: update buildmode pie condition

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented on daemon/logger/gcplogs/gcplogging.go:
```diff
@@ -117,15 +117,6 @@ func New(info logger.Info) (logger.Logger, error) {
-	// So we forcibly set HOME so as to avoid call to os/user/Current()
-	if err := ensureHomeIfStatic(); err != nil {
-		return nil, err
-	}
```
This change is slightly unrelated and can be done in follow-up. (cc @AkihiroSuda)

crazy-max replied:
Removed this commit

crazy-max committed:
Dockerfile: add "all" stage to build binaries and extra tools

We still need a stage that build binaries and extra tools as well for
docker-ce-packaging repo: https://github.com/docker/docker-ce-packaging/blob/ff110508ffce04a9246ef52af4e7a458422e941a/static/Makefile#L41-L57

This could be removed if we create a package for each project
like it's done in docker-packaging repo: https://github.com/docker/packaging/tree/main/pkg

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
Opened https://github.com/docker/docker-ce-packaging/pull/801 to check if packaging is still working with this PR.

corhere commented on pkg/linkmode/dynamic.go:
```diff
@@ -0,0 +1,8 @@
+//go:build !static_build
+// +build !static_build
+
+package linkmode
```
This package does not belong in the `pkg` tree as it makes no sense to be imported by third parties.

corhere commented on pkg/linkmode/dynamic.go:
```diff
@@ -0,0 +1,8 @@
+
+package linkmode
+
+func IsStatic() bool {
+	return false
+}
```
```suggestion
const IsStatic = false
```

corhere commented on pkg/linkmode/static.go:
```diff
@@ -0,0 +1,8 @@
+
+package linkmode
+
+func IsStatic() bool {
+	return true
+}
```
```suggestion
const IsStatic = true
```

corhere commented on hack/make/.binary:
```diff
@@ -73,15 +73,17 @@ source "${MAKEDIR}/.go-autogen"
-			BUILDFLAGS+=("-buildmode=pie")
-			;;
-	esac
+	if [ "$DOCKER_STATIC" = "0" ]; then
```
What happens if `DOCKER_STATIC=trollface`?
```suggestion
	if ! [ "$DOCKER_STATIC" = "1" ]; then
```

corhere commented on hack/make.sh:
```diff
@@ -39,6 +39,14 @@ DEFAULT_BUNDLES=(
+if [[ $VERSION == refs/tags/* ]]; then
+	VERSION=${VERSION#refs/tags/}
+elif [[ $VERSION == refs/heads/* ]]; then
+	VERSION=$(echo "${VERSION#refs/heads/}" | sed -r 's#/+#-#g')
```
You use a here string literally two lines down!
```suggestion
	VERSION=$(<<< "${VERSION#refs/heads/}" sed -r 's#/+#-#g')
```

corhere reviewed (commented):
Looking great!

thaJeztah commented on dockerversion/version_lib.go:
```diff
@@ -6,7 +6,6 @@ var (
 	GitCommit             = "library-import"
 	Version               = "library-import"
 	BuildTime             = "library-import"
-	IAmStatic             = "library-import"
```
ISTR we kept this field in `dockerversion` with the intent to (potentially) surface this information in `docker version` / `docker info` and/or `--version` output to help debugging issues when users report issues (it would save us to ask "is this a static binary, or did you install from deb/rpm?").

I don't think we ever did this, but perhaps still something to consider.

thaJeztah commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1672,7 +1672,7 @@ func (devices *DeviceSet) initDevmapper(doInit bool) (retErr error) {
 	// https://github.com/docker/docker/issues/4036
 	if supported := devicemapper.UdevSetSyncSupport(true); !supported {
-		if dockerversion.IAmStatic == "true" {
+		if linkmode.IsStatic {
```
So, with my other comment in mind; I wonder if we actually _need_ the `linkmode` package; basically, we're passing a built-tag through a `const`; if this is now the only location we use it, and this is only to distinguish between "not supported because.. not supported" or "not supported because static".

Perhaps we should consider;

- changing `UdevSetSyncSupport` to return an error (instead of a `bool`), or to add a utility for that in `pkg/devicemapper`
- `pkg/devicemapper` is already using the `static_build` build-tag to implement different variations, so alternatively we could have some const there.
- Or just remove this code, and make the error message slightly less specific; the device mapper driver is already deprecated, so perhaps that's fine as well.

thaJeztah replied:
- 👆 opened https://github.com/moby/moby/pull/44723 for this one

thaJeztah reviewed (commented):
Haven't done a full re-review yet, but looked at the "static build tag" commit, and left some comments on that; I'll open a PR to remove that part I mentioned.

thaJeztah commented on daemon/logger/gcplogs/gcplogging.go:
```diff
@@ -122,7 +122,7 @@ func New(info logger.Info) (logger.Logger, error) {
 	// in a short term. (golang/go#13470, https://sourceware.org/bugzilla/show_bug.cgi?id=19341)
 	// So we forcibly set HOME so as to avoid call to os/user/Current()
-	if err := ensureHomeIfIAmStatic(); err != nil {
+	if err := ensureHomeIfStatic(); err != nil {
```
- 👆 I opened https://github.com/moby/moby/pull/44722

crazy-max replied:
Yes see my previous comment https://github.com/moby/moby/pull/44546#discussion_r1057682813

thaJeztah commented on hack/make/.binary:
```diff
@@ -73,15 +73,17 @@ source "${MAKEDIR}/.go-autogen"
+			windows/* | linux/mips* | linux/riscv* | linux/ppc64) ;;
+			# TODO remove windows in Go 1.15+: https://github.com/golang/go/commit/95f382139043059a2a0780ba577b53893408f7e4
+			# TODO remove riscv64 in Go 1.16+: https://github.com/golang/go/commit/8eb846fd37eb7bded8a1cf6932be2c59069863e5
+	
```
Was about  to comment: "perhaps fix these", but I see you're doing in the next commit; perhaps would've been slightly less code-churn if done in the reverse order, but "meh".

thaJeztah commented on hack/make/dynbinary-proxy:
```diff
@@ -8,6 +8,7 @@ set -e
 	export BUILDFLAGS=("${BUILDFLAGS[@]/osusergo /}")     # ditto for osusergo
 	export BUILDFLAGS=("${BUILDFLAGS[@]/static_build /}") # we're not building a "static" binary here
 
+	DOCKER_STATIC=0
```
Guess this is redundant, or do we explicitly check for `if DOCKER_STATIC = 0` anywhere?

thaJeztah commented on hack/make/.binary:
```diff
@@ -71,14 +71,11 @@ source "${MAKEDIR}/.go-autogen"
 
-	# -buildmode=pie is not supported on Windows and Linux on mips, riscv64 and ppc64be.
-	# https://github.com/golang/go/blob/77aa209b386a184e7f4b44938f2a05a1b5c5a3cf/src/cmd/internal/sys/supported.go#L89-L99
+	# -buildmode=pie is not supported on Windows arm64 and Linux mips*, ppc64be
+	# https://github.com/golang/go/blob/go1.19.4/src/cmd/internal/sys/supported.go#L125-L132
```
Wondering if this check is one of the things missing in containerd's makefiles / scripts 🤔

thaJeztah commented on docker-bake.hcl:
```diff
@@ -14,8 +59,17 @@ function "bindir" {
 target "_common" {
   args = {
     BUILDKIT_CONTEXT_KEEP_GIT_DIR = 1
-    APT_MIRROR = "cdn-fastly.deb.debian.org"
```
Looks like this commit changes the default from `cdn-fastly.deb.debian.org` to `deb.debian.org`, but doesn't mention that; 

- was there a specific reason to change it? (either one should probably be fine)
- the Dockerfile itself also has handling for defaults (although that part could probably use some improvements to _not make changes_ if no custom value was set)

Would it work to "not" define a value, so that we keep the default central in the Dockerfile?

crazy-max replied:
My bad, that was not intentional. I wonder if defining an alternative mirror is still necessary though.

thaJeztah replied:
We should indeed look if we still want to set this mirror. The option itself _could_ still be useful for users though to pick an alternative (thinking of people behind the GFW)

thaJeztah commented on docker-bake.hcl:
```diff
@@ -1,6 +1,51 @@
+  default = "HEAD"
+}
+
+# Docker version such as 17.04.0-dev. Automatically generated through Git ref.
```
Perhaps we should update the example value to a "non-CalVer" example now 😅

thaJeztah commented on docker-bake.hcl:
```diff
@@ -64,8 +76,8 @@ target "_common" {
     DOCKER_BUILDTAGS = DOCKER_BUILDTAGS
-    DOCKER_GITCOMMIT = DOCKER_GITCOMMIT
-    VERSION = VERSION
+    DOCKER_GITCOMMIT = GITHUB_SHA != "" ? GITHUB_SHA : DOCKER_GITCOMMIT
+    VERSION = GITHUB_REF != "" ? GITHUB_REF : VERSION
```
If feels like the order of preference should be reversed here; `DOCKER_GITCOMMIT` and `VERSION` should (probably?) take precedence if both exist.

thaJeztah commented on docker-bake.hcl:
```diff
@@ -128,19 +128,16 @@ target "binary-cross" {
 # dev
 #
 
-variable "DEV_IMAGE" {
```
Wondering if this was used anywhere (but could've been in CI previously to tie the image name to PR number 🤔). Guess not really important.

crazy-max committed:
Dockerfile: use TARGETPLATFORM to build Docker

Better support for cross compilation so we can fully rely
on `--platform` flag of buildx for a seamless integration.

This removes unnecessary extra cross logic in the Dockerfile,
DOCKER_CROSSPLATFORMS and CROSS vars and some hack scripts as well.

Non-sandboxed build invocation is still supported and dev stages
in the Dockerfile have been updated accordingly.

Bake definition and GitHub Actions workflows have been updated
accordingly as well.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
hack: remove containerutility script

this script is not used anymore. containerutility is
built in the Dockerfile.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
bake: define additional vars

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
Thanks! @corhere approved this PR earlier, and changes since his last review were removing some changes (related to https://github.com/moby/moby/pull/44722).

I asked @rumpl to do another review, to have the extra eyes before I merge; let's get this one in 👍

thaJeztah commented:
oh! didn't expect you to  be online, @corhere 😂  thanks!
